### PR TITLE
Integration/realtek sdw codecs

### PR DESCRIPTION
**Pull request by plbossart**
PR to track Realtek codecs. This will be removed once the code is all upstream.

### CODE REVIEW

**Comment by plbossart** [comment]
@shumingfan @bardliao @RanderWang this is FYI to track updates

File: sound/soc/codecs/rt711-sdw.c [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+	.bus_config = rt711_bus_config,
+};
+
+static int rt711_sdw_probe(struct sdw_slave *slave,
+			   const struct sdw_device_id *id)
+{
+	struct regmap *regmap;
+	int ret = 0;
+
+	pr_err("bard: %s\n", __func__);
```
Comment (RanderWang) [inline sound/soc/codecs/rt711-sdw.c]: bard ?

**Comment by RanderWang** [comment]
@shumingfan I have a suggestion: please use dev_dbg instead of pr_debug.  three reasons:
(1) dev_dbg is more popular in ASoC/codec/
(2) dev_dbg prints dev info which is easy for debugging with multi-codecs
(3) Asoc maintainer prefers dev_dbg according to bard

**Comment by shumingfan** [comment]
> @shumingfan I have a suggestion: please use dev_dbg instead of pr_debug. three reasons:
> (1) dev_dbg is more popular in ASoC/codec/
> (2) dev_dbg prints dev info which is easy for debugging with multi-codecs
> (3) maintainer prefers dev_dbg according to bard

Yeah, I cleaned up the codec driver and won't use pr_debug.

**Comment by shumingfan** [comment]
Hi all,
I had submitted a PR #1144. Please take a look.

**Comment by bardliao** [comment]
@plbossart Do we need to close this one and review on #1144 ?

**Comment by plbossart** [comment]
I will take Realtek's patches, squash them and push a new version here, that way we can directly review all the code in one shot.

**Comment by plbossart** [comment]
Realtek patches squashed, authors and commit messages updated. Same functionality as PR1144, only self contained in one patch per codec.

**Comment by plbossart** [comment]
PR updated, now on top of 20190821 topic/sof-dev-rebase

File: sound/soc/codecs/rt715.c [hunk 2]
```diff
@@ -0,0 +1,1274 @@
...
+	rt715->hw_init = false;
+
+	/*
+	 * TODO: The BIOS used currently has multiple entries
+	 * for Slave(s). Ideally for a given platform, BIOS should
+	 * hold entries of actual Slave(s) present on board. So
+	 * to handle this, only register codec if it is actually
+	 * present on board, adding check accordingly
+	 */
+	if (slave->bus->link_id == 2) {
```
Comment (bardliao) [inline sound/soc/codecs/rt715.c]: @jack-cy-yu Please remove the if condition.  rt715 may connect to any sdw link

File: sound/soc/codecs/rt700-sdw.c [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+};
+
+static int rt700_update_status(struct sdw_slave *slave,
+			       enum sdw_slave_status status)
+{
+	struct rt700_priv *rt700 = dev_get_drvdata(&slave->dev);
+
+	/* Update the status */
+	rt700->status = status;
+
```
Comment (bardliao) [inline sound/soc/codecs/rt700-sdw.c:318]: Should we set ``rt700->hw_init = false`` if ``rt700->status == SDW_SLAVE_UNATTACHED``?

File: sound/soc/codecs/rt1308-sdw.c [hunk 4]
```diff
@@ -0,0 +1,688 @@
...
+		value = 0x5;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	regmap_write(rt1308->regmap, 0xe0, value);
+	regmap_write(rt1308->regmap, 0xf0, value);
+
+	dev_dbg(dev, "%s clock_config complete\n", __func__);
```
Comment (bardliao) [inline sound/soc/codecs/rt1308-sdw.c]: Maybe print clk_freq here will be useful.

File: sound/soc/codecs/rt1308-sdw.c [hunk 5]
```diff
@@ -0,0 +1,688 @@
...
+	struct rt1308_sdw_priv *rt1308 = dev_get_drvdata(&slave->dev);
+	int ret;
+
+	memcpy(&rt1308->params, params, sizeof(*params));
+
+	ret = rt1308_clock_config(&slave->dev);
+	if (ret < 0)
+		dev_err(&slave->dev, "Invalid clk config");
+
+	return 0;
```
Comment (bardliao) [inline sound/soc/codecs/rt1308-sdw.c]: return ret?

File: sound/soc/codecs/rt1308-sdw.c [hunk 6]
```diff
@@ -0,0 +1,688 @@
...
+	if (!regmap)
+		return -EINVAL;
+
+	rt1308_sdw_init(&slave->dev, regmap, slave);
+
+	/* Perform IO operations only if slave is in ATTACHED state */
+	if (slave->status == SDW_SLAVE_ATTACHED)
+		rt1308_io_init(&slave->dev, slave);
+
+	return ret;
```
Comment (bardliao) [inline sound/soc/codecs/rt1308-sdw.c]: It seems that we don't use _ret_? Should we just return 0?

File: sound/soc/codecs/rt711-sdw.c [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+	struct rt711_priv *rt711 = dev_get_drvdata(&slave->dev);
+	int ret;
+
+	memcpy(&rt711->params, params, sizeof(*params));
+
+	ret = rt711_clock_config(&slave->dev);
+	if (ret < 0)
+		dev_err(&slave->dev, "Invalid clk config");
+
+	return 0;
```
Comment (bardliao) [inline sound/soc/codecs/rt711-sdw.c]: return ret?

File: sound/soc/codecs/rt711.c [hunk 8]
```diff
@@ -0,0 +1,1232 @@
...
+	if (ret < 0) {
+		pr_err("Failed to set private value: %d\n", ret);
+		goto err;
+	}
+	ret = regmap_write(regmap, RT711_PRIV_DATA_W_L | nid, val_l);
+	if (ret < 0) {
+		pr_err("Failed to set private value: %d\n", ret);
+		goto err;
+	}
+	return 0;
```
Comment (bardliao) [inline sound/soc/codecs/rt711.c]: Should we remove the ``return 0;`` and let it return ret?

File: sound/soc/codecs/rt711.c [hunk 9]
```diff
@@ -0,0 +1,1232 @@
...
+	int ret;
+
+	rt711->component = component;
+
+	/* calibration */
+	ret = rt711_calibration(rt711);
+	if (ret < 0)
+		dev_err(component->dev, "%s calibration failed\n", __func__);
+
+	return 0;
```
Comment (bardliao) [inline sound/soc/codecs/rt711.c:890]: return ret?

File: sound/soc/codecs/rt711.c [hunk 10]
```diff
@@ -0,0 +1,1232 @@
...
+	unsigned int val = 0;
+
+	dev_dbg(dai->dev, "%s %s", __func__, dai->name);
+	stream = snd_soc_dai_get_dma_data(dai, substream);
+
+	if (!stream)
+		return -ENOMEM;
+
+	if (!rt711->slave)
+		return 0;
```
Comment (bardliao) [inline sound/soc/codecs/rt711.c]: Shouldn't this be an error?

File: sound/soc/codecs/rt711.c [hunk 11]
```diff
@@ -0,0 +1,1232 @@
...
+int rt711_pcm_hw_free(struct snd_pcm_substream *substream,
+		      struct snd_soc_dai *dai)
+{
+	struct snd_soc_component *component = dai->component;
+	struct rt711_priv *rt711 = snd_soc_component_get_drvdata(component);
+	struct sdw_stream_data *stream =
+		snd_soc_dai_get_dma_data(dai, substream);
+
+	if (!rt711->slave)
+		return 0;
```
Comment (bardliao) [inline sound/soc/codecs/rt711.c]: return -EINVAL?

File: sound/soc/codecs/rt715-sdw.c [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+		dev_err(&slave->dev, "Invalid clk config");
+
+	return 0;
+}
+
+static struct sdw_slave_ops rt715_slave_ops = {
+	.read_prop = rt715_read_prop,
+	.update_status = rt715_update_status,
+	.bus_config = rt715_bus_config,
+	.port_prep = NULL,
```
Comment (bardliao) [inline sound/soc/codecs/rt715-sdw.c]: Won't it be NULL if we don't set it?

File: sound/soc/codecs/rt715-sdw.c [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+	if (!regmap)
+		return -EINVAL;
+
+	rt715_init(&slave->dev, regmap, slave);
+
+	/* Perform IO operations only if slave is in ATTACHED state */
+	if (slave->status == SDW_SLAVE_ATTACHED)
+		rt715_io_init(&slave->dev, slave);
+
+	return ret;
```
Comment (bardliao) [inline sound/soc/codecs/rt715-sdw.c]: remove ret and return 0?

File: sound/soc/codecs/rt715.c [hunk 14]
```diff
@@ -0,0 +1,1270 @@
...
+/*
+ * rt715.c -- rt715 ALSA SoC audio driver
+ *
+ * Copyright(c) 2019 Realtek Semiconductor Corp.
+ *
+ * ALC715 ASoC Codec Driver based Intel Dummy SdW codec driver
+ *
+ */
+
+// #define DEBUG
```
Comment (bardliao) [inline sound/soc/codecs/rt715.c]: remove this if the driver is for upstream.

File: sound/soc/codecs/rt715.c [hunk 15]
```diff
@@ -0,0 +1,1270 @@
...
+	if (ret < 0) {
+		pr_err("Failed to set private value: %d\n", ret);
+		goto err;
+	}
+	ret = regmap_write(regmap, RT715_PRIV_DATA_W_L, val_l);
+	if (ret < 0) {
+		pr_err("Failed to set private value: %d\n", ret);
+		goto err;
+	}
+	return 0;
```
Comment (bardliao) [inline sound/soc/codecs/rt715.c]: Can we remove the ``return 0;``?

File: sound/soc/codecs/rt715.c [hunk 16]
```diff
@@ -0,0 +1,1270 @@
...
+	"MIC1",
+	"MIC1",
+	"DMIC1",
+	"DMIC2",
+	"DMIC3",
+	"DMIC4",
+};
+
+static const SOC_ENUM_SINGLE_DECL(
+	rt715_adc22_enum, RT715_MUX_IN1, 0, adc_22_23_mux_text);
```
Comment (bardliao) [inline sound/soc/codecs/rt715.c:399]: I am not sure will the reg (RT715_MUX_IN1) be set by DAPM when the Mux is power up/down.

File: sound/soc/codecs/rt715.c [hunk 17]
```diff
@@ -0,0 +1,1270 @@
...
+int rt715_pcm_hw_free(struct snd_pcm_substream *substream,
+		      struct snd_soc_dai *dai)
+{
+	struct snd_soc_component *component = dai->component;
+	struct rt715_priv *rt715 = snd_soc_component_get_drvdata(component);
+	struct sdw_stream_data *stream =
+		snd_soc_dai_get_dma_data(dai, substream);
+
+	if (!rt715->slave)
+		return 0;
```
Comment (bardliao) [inline sound/soc/codecs/rt715.c]: return -EINVAL;?

File: sound/soc/codecs/rt715.c [hunk 18]
```diff
@@ -0,0 +1,1270 @@
...
+	int retval, port, num_channels;
+	unsigned int val = 0;
+
+	stream = snd_soc_dai_get_dma_data(dai, substream);
+
+	if (!stream)
+		return -ENOMEM;
+
+	if (!rt715->slave)
+		return 0;
```
Comment (bardliao) [inline sound/soc/codecs/rt715.c]: return -EINVAL;?

File: sound/soc/codecs/rt700-sdw.c [hunk 19]
```diff
@@ -298,6 +298,13 @@ static int rt700_sdw_probe(struct sdw_slave *slave,
 
 static int rt700_sdw_remove(struct sdw_slave *slave)
 {
+	struct rt700_priv *rt700 = dev_get_drvdata(&slave->dev);
+
+	if (rt700 && rt700->hw_init) {
+		cancel_delayed_work(&rt700->jack_detect_work);
+		cancel_delayed_work(&rt700->jack_btn_check_work);
+	}
+
```
Comment (bardliao) [inline sound/soc/codecs/rt700-sdw.c:479]: Should we set ``rt700->hw_init = false;`` here?

**Comment by shumingfan** [comment]
> @shumingfan Here are some comments from my side. I may not write comments for all codec driver if there is a similar issue in different codec drivers. So, please check all other codec drivers when you see a comment on any codec drivers.

Thanks, I will check all you comments to modify the driver. I will send the patches later.

File: sound/soc/codecs/rt700-sdw.c (hunk 19 above)
Comment (plbossart) [reply sound/soc/codecs/rt700-sdw.c:479]: not sure if this is needed, the rt700 structure will be freed on remove (was allocated with devm_kzalloc())

**Comment by plbossart** [comment]
merge/squashed patches and added minor style corrections (removed extra newlines, balanced brackets) plus one typo. please rebase for any follow-up corrections

**Comment by plbossart** [comment]
branch rebased on top of topic/sof-dev to start triggering builds
This does not yet test SoundWire functionality in CI but at least that code builds and does not harm

**Pull request closed without merging**